code/parachain/frame/assets-interface/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+
+	/// Asset Nonce
+	#[pallet::storage]
+	#[allow(clippy::disallowed_types)] // Allow for `ValueQuery` because of nonce
+	pub type AssetNonce<T: Config> =
+		StorageValue<_, u64, ValueQuery, Nonce<ZeroInit, SafeIncrement>>;
+
+	/// Fee to pay for new asset registry
+	#[pallet::storage]
+	pub type AssetCreationFee<T: Config> = StorageValue<_, T::Balance, OptionQuery>;
+
+	/// Asset admin
+	#[pallet::storage]
+	pub type AssetAdmin<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, T::AccountId, OptionQuery>;
+
+	/// Is Asset CosmWasm Asset
+	#[pallet::storage]
+	pub type IsCosmWasmAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, (), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NewCreationFeeSet(T::Balance),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetNotFound,
+		NotAssetAdmin,
+		BadDecimals,
+		TransferError,
+		CreationFeeIsNotSet,
+		NotCosmWasmAsset,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates an asset for bridge transfer
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::register_asset())]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			location: T::Location,
+			name: Option<BiBoundedAssetName>,
+			symbol: Option<BiBoundedAssetSymbol>,
+			decimals: u8,
+			is_cosmwasm: bool,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let protocol_id = (Pallet::<T>::index() as u32).to_be_bytes();
+
+			let asset_id = <T::AssetsRegistry as GenerateAssetId>::generate_asset_id(
+				protocol_id,
+				AssetNonce::<T>::increment().expect("Does not exceed u64::MAX"),
+			);
+			<Self as PermissionlessAsset>::register_asset(
+				who,
+				asset_id,
+				location,
+				name,
+				symbol,
+				decimals,
+				is_cosmwasm,
+			)?;
+			Ok(())
+		}
+
+		/// Creates an asset for bridge transfer
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_creation_fee())]
+		pub fn set_creation_fee(origin: OriginFor<T>, fee_amount: T::Balance) -> DispatchResult {
+			T::SetFeeOrigin::ensure_origin(origin)?;
+			AssetCreationFee::<T>::put(fee_amount);
+			Self::deposit_event(Event::NewCreationFeeSet(fee_amount));
+			Ok(())
+		}
+
+		/// Update stored asset information.
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as Config>::WeightInfo::update_asset())]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			name: UpdateValue<Option<BiBoundedAssetName>>,
+			symbol: UpdateValue<Option<BiBoundedAssetSymbol>>,
+			decimals: UpdateValue<Option<u8>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let asset_admin =
+				AssetAdmin::<T>::try_get(asset_id).map_err(|_| Error::<T>::AssetNotFound)?;
+			ensure!(asset_admin == who, Error::<T>::NotAssetAdmin);
+			if let UpdateValue::Set(Some(decimals_update)) = decimals {
+				ensure!(decimals_update > 0 && decimals_update < 30, Error::<T>::BadDecimals);
```

Comment:
do not check here. it would be checked inside AR

Reply:
we allow decimals be None in AR

Reply:
i am ok with none everywhere. Coin in cw does not have decimals too

Reply:
George said it is part of token creation

Reply:
they will use it, fine do them. why making it non optional and do some checks?

Reply:
i see what you mean but we use decimals for rate limiter logic rn so maybe we shouldnt allow them to input anything they want

Reply:
decimal in rate limiter are not used. it is not real. think about.

Reply:
i know it isnt real but as of rn we use it https://github.com/ComposableFi/composable/blob/main/code/parachain/runtime/picasso/src/ibc.rs#L109

Reply:
idiotic non sence

Reply:
plug 10 to 13 and nothing really changes, or any other